quic/s2n-quic/src/provider/event/console_perf.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::provider::event;
+use core::time::Duration;
+use std::sync::{
+    atomic::{AtomicU64, Ordering},
+    Arc,
+};
+
+/// An event subscriber that prints performance metrics to the console
+///
+/// # Examples
+///
+/// Enables the console perf event subscriber for the server,
+/// configured to print metrics every 10 seconds
+///
+/// ```rust,ignore
+/// # use std::{error::Error, time::Duration};
+/// use s2n_quic::Server;
+/// #
+/// # #[tokio::main]
+/// # async fn main() -> Result<(), Box<dyn Error>> {
+/// let server = Server::builder()
+///     .with_event(event::console_perf::Subscriber::new(core::time::Duration::from_secs(10)))?
+///     .start()?;
+/// #
+/// #    Ok(())
+/// # }
+/// ```
+#[derive(Debug, Default)]
+pub struct Subscriber {
+    pub counters: Arc<Counters>,
+    pub frequency: Duration,
+    pub spawned: bool,
+}
+
+impl Subscriber {
+    /// Create a new `console_perf::Subscriber` that
+    /// prints metrics at the given `frequency`
+    pub fn new(frequency: Duration) -> Self {
+        Self {
+            counters: Default::default(),
+            frequency,
+            spawned: false,
+        }
+    }
+
+    /// Create a new `console_perf::Subscriber` that
+    /// does not print to the console
+    pub fn disabled() -> Self {
+        Self {
+            counters: Default::default(),
+            frequency: Duration::MAX,
+            spawned: true,
+        }
+    }
+
+    fn spawn(&mut self) {
+        if !self.spawned {
+            self.spawned = true;
+            let counters = self.counters.clone();
+            let frequency = self.frequency;
+            counters.print_header();
+            tokio::spawn(async move {
+                loop {
+                    tokio::time::sleep(frequency).await;
+                    counters.print(frequency);
+                }
+            });
+        }
+    }
```

Comment:
I wonder if it would be better to just have a `print` method and then the caller can decide how to spawn a task (e.g. tokio, system threads, etc.) or how often to call it.

Reply:
Yeah, there's a tradeoff between flexibility and ease of use. Maybe I could make it so if the user supplies a frequency it auto spawns, otherwise the user can handle spawning

Reply:
I think the other issue is this assumes `tokio` in the environment, which may not always be the case long term. If we let the application handle the spawning it's up to them to figure out how to do the spawn/sleep

Reply:
Having some trouble seeing how would work, since the caller would be doing something like:
```rust
subscriber.print_header();
let frequency = Duration::from_secs(1);
tokio::spawn(async move {
    loop {
            tokio::time::sleep(frequency).await;
            subscriber.print(frequency);
    });
```
but this would cause borrow issues since the subscriber is also passed to the client/server:
```rust
error[E0382]: use of moved value: `subscriber`
   --> quic/s2n-quic-qns/src/client/perf.rs:133:27
    |
116 |           let mut subscriber = event::console_perf::Builder::default()
    |               -------------- move occurs because `subscriber` has type `s2n_quic::provider::event::console_perf::Subscriber`, which does not implement the `Copy` trait
...
123 |               tokio::spawn(async move {
    |  _____________________________________-
124 | |                 loop {
125 | |                     tokio::time::sleep(frequency).await;
126 | |                     subscriber.print(frequency);
    | |                     ---------- variable moved due to use in generator
127 | |                 }
128 | |             });
    | |_____________- value moved here
...
133 |           let subscriber = (subscriber, event::tracing::Subscriber::default());
    |                             ^^^^^^^^^^ value used here after move
```

Reply:
yeah it would need to be

```rust
let mut subscriber = event::console_perf::Builder::default();
tokio::spawn({
    let subscriber = subscriber.clone();
    async move {
        subscriber.print_header();
        loop {
            tokio::time::sleep(freq).await;
            subscriber.print();
        }
    }
});
```

Reply:
ah yep, I didn't have subscriber implementing `Clone`